internal/controller/logger_manager.go

```diff
@@ -0,0 +1,267 @@
+			LogEntry: &pb.LogEntry{
+				Level:       pb.LogLevel(logEntry.Level),
+				Time:        timestamppb.New(logEntry.Time),
+				JsonMessage: buf.String(),
+			},
+		},
+	})
+	return err
+}
+
+// UpdateClient will update BufferedGrpcWriteSyncer with new client stream and GRPC connection
+func (b *BufferedGrpcWriteSyncer) UpdateClient(client pb.KubernetesInfoService_SendLogsClient, conn *grpc.ClientConn) {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+	b.client = client
+	b.conn = conn
+
+	// If logs were lost, send a log message about it
+	if b.conn.GetState() == connectivity.Ready && b.lostLogs {
+		lostLogMessage := zapcore.Entry{
+			Level:   zapcore.ErrorLevel,
+			Time:    time.Now(),
+			Message: b.lostLogMessage(),
+		}
+		err := b.sendLog(lostLogMessage)
+		if err != nil {
+			b.logger.Warn("Error while sending log",
+				"error", err,
+			)
+		}
+		b.lostLogs = false
+		b.lostBytes = 0
+	}
+}
+
+// ListenToLogStream will wait for responses from server and will update log level
+// depending respone contents
+func (b *BufferedGrpcWriteSyncer) ListenToLogStream() {
+	for {
+		res, err := b.client.Recv()
+		if err == io.EOF {
+			// The client has closed the stream
+			b.logger.Info("Server has closed the SendLogs stream")
+			return
+		}
+		if err != nil {
+			b.logger.Errorw("Stream terminated",
+				"error", err,
+			)
+			return
+		}
+		switch res.Response.(type) {
+		case *pb.SendLogsResponse_SetLogLevel:
+			newLevel := res.GetSetLogLevel().Level
+			b.updateLogLevel(newLevel)
+		}
+	}
+}
+
+// bufferLog adds the log entry to in-memory buffer
+func (b *BufferedGrpcWriteSyncer) bufferLog(entry zapcore.Entry) {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+
+	b.buffer = append(b.buffer, entry)
+	if len(b.buffer) >= maxBufferSize {
+		b.flush()
+	}
+}
+
+// lostLogMessage generates a message about lost logs
+func (b *BufferedGrpcWriteSyncer) lostLogMessage() string {
+	return "Lost logs due to buffer overflow. Total lost bytes: " + strconv.Itoa(b.lostBytes)
+}
+
+// updateLogLevel sets the logger's log level based on the response from the server.
+func (b *BufferedGrpcWriteSyncer) updateLogLevel(level pb.LogLevel) {
+	switch level {
+	case pb.LogLevel_LOG_LEVEL_DEBUG:
+		b.logger.Info("Set to DEBUG level log")
+		b.logLevel.SetLevel(zapcore.DebugLevel)
+	case pb.LogLevel_LOG_LEVEL_ERROR:
+		b.logger.Info("Set to ERROR level log")
+		b.logLevel.SetLevel(zapcore.ErrorLevel)
+	case pb.LogLevel_LOG_LEVEL_INFO:
+		b.logger.Info("Set to INFO level log")
+		b.logLevel.SetLevel(zapcore.InfoLevel)
+	case pb.LogLevel_LOG_LEVEL_WARN:
+		b.logger.Info("Set to WARN level log")
+		b.logLevel.SetLevel(zapcore.WarnLevel)
+	default:
+		b.logger.Warn("Unknown log level received, defaulting to INFO")
+		b.logLevel.SetLevel(zapcore.InfoLevel)
+	}
+}
+
+// NewGrpclogger will define a new zap logger with multiple writesyncs
+// one to stdout and one for GRPC writestream
+func NewGrpclogger(grpcSyncer *BufferedGrpcWriteSyncer) *zap.SugaredLogger {
```

Comment:
By convention, acronyms are CAPITALIZED in Golang symbols.
```suggestion
func NewGRPClogger(grpcSyncer *BufferedGrpcWriteSyncer) *zap.SugaredLogger {
```